File: app/src/main/java/aleksandrkim/ArchComponentsTest/NoteDetails/NoteDetailsFragment.java
package aleksandrkim.ArchComponentsTest.NoteDetails;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import aleksandrkim.ArchComponentsTest.HostActivity.NavigationActivity;
import aleksandrkim.ArchComponentsTest.R;
import aleksandrkim.ArchComponentsTest.Utils.Colors;

public class NoteDetailsFragment extends Fragment implements NavigationActivity.BackEnabled {
    public static final String TAG = "ComposeFragment";
    private static final String KEY_NOTE_ID = "noteIdParam";
    private static final String KEY_NOTE_COLOR = "noteColor";
    private int noteId;
    private NavigationActivity navigationActivity;

    EditText etTitle;
    EditText etContent;

    private NoteDetailsVM noteDetailsViewModel;

    public NoteDetailsFragment() {}

    public static NoteDetailsFragment newInstance(@Nullable int noteId) {
        NoteDetailsFragment fragment = new NoteDetailsFragment();

        Bundle args = new Bundle();
        args.putInt(KEY_NOTE_ID, noteId);
        fragment.setArguments(args);
        return fragment;
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);

        if (getArguments() != null)
            this.noteId = getArguments().getInt(KEY_NOTE_ID, -1);

        initVM(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_note_compose, container, false);
        etTitle = v.findViewById(R.id.et_title);
        etContent = v.findViewById(R.id.et_content);

        if (savedInstanceState == null) setEt();

        return v;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        navigationActivity = (NavigationActivity) requireActivity();
        navigationActivity.setTitle(R.string.new_note);
        navigationActivity.setUpButton(true);
    }

    private void saveNote() {
        updateCurrentNote();
        if (!noteDetailsViewModel.hasEitherField()) {
            Toast.makeText(getActivity(), getString(R.string.cannot_save_empty_note), Toast.LENGTH_SHORT).show();
            return;
        }
        noteDetailsViewModel.addOrUpdateCurrentNote();
    }

    private void changeColorTag() {
        AlertDialog.Builder builder = new AlertDialog.Builder((Context) navigationActivity);
        builder.setTitle(R.string.choose_color)
                .setItems(Colors.colorTitles, (dialog, which) -> noteDetailsViewModel.setColor(Colors.colors[which]));
        builder.create().show();
    }

    private void initVM(Bundle savedInstanceState){
        noteDetailsViewModel = ViewModelProviders.of(this).get(NoteDetailsVM.class);
        noteDetailsViewModel.setCurrentNote(noteId, savedInstanceState != null ? savedInstanceState.getInt(KEY_NOTE_COLOR) : null);
    }

    private void setEt() {
        etTitle.setText(noteDetailsViewModel.getTitle());
        etContent.setText(noteDetailsViewModel.getContent());
    }

    private void updateCurrentNote() {
        noteDetailsViewModel.setTitle(etTitle.getText().toString());
        noteDetailsViewModel.setContent(etContent.getText().toString());
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.note_compose_menu, menu);
        noteDetailsViewModel.getColor().observe(this, integer -> menu.getItem(0).getIcon().setTint(integer));
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                saveNote();
                return false;
            case R.id.menu_color_pick:
                changeColorTag();
                return true;
        }
        return false;
    }

    @Override
    public void onBackPressed() {
        saveNote();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        Log.d(TAG, "onSaveInstanceState: ");
        outState.putInt(KEY_NOTE_COLOR, noteDetailsViewModel.getColor().getValue());
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroyView() {
        noteDetailsViewModel.removeAllObs(this);
        super.onDestroyView();
    }
}

File: app/src/main/java/aleksandrkim/MinimalNotes/NoteDetails/NoteDetailsFragment.kt
package aleksandrkim.MinimalNotes.NoteDetails

import aleksandrkim.MinimalNotes.HostActivity.NavigationActivity
import aleksandrkim.MinimalNotes.R
import aleksandrkim.MinimalNotes.Utils.Colors
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AlertDialog
import android.view.*
import android.widget.Toast
import kotlinx.android.synthetic.main.fragment_note_details.*

class NoteDetailsFragment : Fragment(), NavigationActivity.BackEnabled {
    private var noteId: Int = -1

    private lateinit var navigationActivity : NavigationActivity
    private val noteDetailsViewModel: NoteDetailsVM by lazy { ViewModelProviders.of(this).get(NoteDetailsVM::class.java) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setHasOptionsMenu(true)

        arguments?.let { noteId = it.getInt(KEY_NOTE_ID, -1) }

        noteDetailsViewModel.setCurrentNote(noteId, savedInstanceState?.getInt(KEY_NOTE_COLOR))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_note_details, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState == null) observeAvailableNote()
        navigationActivity = requireActivity() as NavigationActivity
        navigationActivity.setTitle(R.string.new_note)
        navigationActivity.setUpButton(true)
    }

    private fun observeAvailableNote() {
        noteDetailsViewModel.currentNote.observe(this, Observer {
            title.setText(it?.title)
            body.setText(it?.body)
            noteDetailsViewModel.color.value = it?.color // set initial color value as soon as it is fetched
        })
    }

    private fun changeColorTag() {
        val builder = AlertDialog.Builder(navigationActivity as Context)
        builder.setTitle(R.string.choose_color)
            .setItems(Colors.colorTitles) { _, which -> noteDetailsViewModel.color.value = Colors.colors[which] }.create().show()
    }

    private fun saveNote() {
        updateCurrentNote()
        if (!noteDetailsViewModel.currentNote.value!!.isBlank()) {
            noteDetailsViewModel.addOrUpdateCurrentNote()
        } else {
            Toast.makeText(activity, getString(R.string.cannot_save_empty_note), Toast.LENGTH_SHORT).show()
        }
    }

    private fun updateCurrentNote() {
        noteDetailsViewModel.currentNote.value?.title = title.text.toString()
        noteDetailsViewModel.currentNote.value?.body = body.text.toString()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        inflater.inflate(R.menu.note_compose_menu, menu)
        noteDetailsViewModel.color.observe(this, Observer { menu.getItem(0).icon.setTint(it ?: -1) })
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                saveNote()
                false
            }
            R.id.menu_color_pick -> {
                changeColorTag()
                true
            }
            else -> false
        }
    }

    override fun onBackPressed() {
        saveNote()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(KEY_NOTE_COLOR, noteDetailsViewModel.color.value ?: -1)
        super.onSaveInstanceState(outState)
    }

    override fun onDestroyView() {
        noteDetailsViewModel.removeAllObs(this)
        super.onDestroyView()
    }

    companion object {
        const val TAG = "NoteDetailsFragment"
        const val KEY_NOTE_ID = "noteIdParam"
        const val KEY_NOTE_COLOR = "noteColor"

        fun newInstance(noteId: Int): NoteDetailsFragment {
            val fragment = NoteDetailsFragment()

            val args = Bundle()
            args.putInt(KEY_NOTE_ID, noteId)
            fragment.arguments = args
            return fragment
        }
    }
}
